[FEAT] 생성 페이지 지도 구현

Da-Yeon-Kim opened the pull request:
## 주요 변경사항 ✔️
Kakao Map API, Local API 등을 이용하여 지도를 구현했습니다.
백에게 장소 ID를 줘야 하는데 Map API만으로는 불가능해서 Local API까지 끌어들였습니다,, ㅎㅎ
조금 로직이 복잡한데 (좌표 👉 주소 👉 ID) 이렇게 가져와야 해요.
- `useGeolocation` : 사용자의 현재 위치를 받아옵니다. (단, 허용 시)
- `useKakaoMap` : 사용자가 클릭하면 지도에 마커를 표시합니다.
- `useGeocoder` : 좌표로부터 주소를 가져옵니다.
- `usePlaceSearch` : 주소로부터 장소 ID를 가져옵니다.

API 명세서에 따라 baseLocation을 POST 요청 데이터에 추가했습니다.
그리고 `생성하기` 버튼 클릭 시, 장소 지정되었는지 확인 후 API 통신합니다!

## 리뷰어에게... 🙏
훅들 사이에서 자주 사용되는 공통 타입이 있어서 `types` 폴더를 따로 생성했습니다.
코드를 다시 점검하질 않아서 좀 더럽거나 중복 작업이 있을 수도 있습니다!
살펴보실 때, 내 현재 위치 가져오기가 잘 되는지 확인 한 번만 부탁드려요!
웹이라 그런지 어째 저는 됐다 말았다 하네요.
(그래도 이게 잘 마쳐지면 생성페이지 드디어 끝!😀)


## 관련 이슈

closes #50

Da-Yeon-Kim committed:
feat(CreateMap): 카카오맵 API 활용해 표시

- 카카오맵 API로 지도 보이기
- geolocation 이용해 현재 위치를 기본 위치로 하기 (단, 위치 허용)
- 지도를 클릭하여 핀 꽂고, 위도/경도 받아오기

 #50

Da-Yeon-Kim committed:
feat(CreateMap): 마커한 장소 정보 데이터 출력

- useGecoder를 이용해 좌표로부터 주소 정보 가져오기
- usePlaceSearch를 이용해 주소 정보로부터 주소 id 가져오기
- type은 중복 사용이 잦아 따로 분리해 관리
- 폴더 구조 고민 중....

 #50

Da-Yeon-Kim committed:
refactor(CreateMap): 훅, 타입 위치 변경

- useGeocoder, useGeolocation, useKakaoMap, usePlaceSearch를 hooks
  폴더로 이동
  - type.ts 파일을 types 폴더로 이동 및 이름 변경

   #50

Da-Yeon-Kim committed:
feat(CreateMap): API 통해 백으로 장소 정보 요청

- 요청 데이터에 baseLocation 추가
- CreateFormInputs에서 title, durationTime으로 변경
(기존 : MeetingName, duration)
- 생성하기 버튼 클릭 시, 장소 선택했는지 확인

 #50

Da-Yeon-Kim committed:
feat(CreateBtn): 장소 정보 입력 검증 분리

- 장소 정보 validation도 분리해 관리
- durationTime number로 반환하도록 변경
- 최종 장소 콘솔에 출력하는 코드 삭제 (개발 때 확인용)

 #50

Da-Yeon-Kim committed:
feat(CreateMap): 지도에 반경 표시

- 일단 임시 범위로 지정 (radius 500)
- 추후 바뀔 수 있음
- 반경이 모두 보이도록 지도 레벨 조정

 #50

geongyu09 commented on src/components/features/Create/CreateBtn/index.tsx:
```diff
@@ -25,10 +25,11 @@ export const CreateBtn: React.FC = () => {
       startDate: values.startDate?.toISOString().split('T')[0] || '',
       endDate: values.endDate?.toISOString().split('T')[0] || '',
-      durationTime: values.duration,
+      durationTime: Number(values.durationTime),
```
`durationTime: +values.durationTime` 로 작성할 수도 있겠네요!

geongyu09 commented on src/components/features/Create/CreateForm/MeetingName/index.tsx:
```diff
@@ -14,7 +14,7 @@ export const MeetingName: React.FC = () => {
         id="meetingName"
         placeholder="모임 이름을 입력해주세요"
-        {...register('meetingName')}
+        {...register('title')}
```
용어 통일한 부분이 정말 좋네요! 👍👍👍👍

geongyu09 commented on src/components/features/Create/CreateBtn/validation.ts:
```diff
@@ -3,7 +3,7 @@ import type { CreateFormInputs } from '@/hooks/useCreateFormContext';
   values: CreateFormInputs,
 ): { errorMessage?: string; isValid: boolean } => {
-  if (!values.meetingName || values.meetingName.trim() === '') {
+  if (!values.title || values.title.trim() === '') {
```
value. 이 게속 큰 의미 없이 반복 되는 것 같아서 아래와 같이 구조분해할당 해서 사용해도 좋아보입니다! 
`const { baseLocation, durationTime, endDate, endTime, startDate, startTime, title } = values;`

geongyu09 commented on src/components/features/Create/CreateMap/index.tsx:
```diff
@@ -1,16 +1,40 @@
+  useEffect(() => {
+    if (placeInfo) {
+      setValue('baseLocation', {
+        location_id: Number(placeInfo.location_id),
```
+placeInfo.location_id, 로 바꿀 수도 있겠네요!

geongyu09 commented on src/components/features/Create/CreateMap/index.tsx:
```diff
@@ -1,16 +1,40 @@
-  return <MapImg />;
+export const CreateMap: React.FC = () => {
+  const { setValue } = useFormContext<CreateFormInputs>();
+  const userLocation: Coordinates | null = useGeolocation();
```
타입을 추론하도록 해도 괜찮아보입니다!

geongyu09 commented on src/hooks/useGeolocation.ts:
```diff
@@ -0,0 +1,29 @@
+          console.error("Error getting the user's location:", error);
+          setLocation({
+            lat: 37.5665,
+            lng: 126.978,
```
혹시 해당 값은 어떤 값인가요? 
값이 어떤 의미를 담고 있다면, 이를 상수로 만들어도 괜찮아보입니다!

Da-Yeon-Kim replied:
아마 API 명세서에 있던 예시 값이든가, 아니면 카카오맵 API 공식문서에 있던 값일 거예요!
딱히 중요한 의미의 값은 아니었는데ㅋㅋㅋ 상수로 만들어서 관리하는 건 좋아 보입니다!

geongyu09 commented on src/hooks/useKakaoMap.ts:
```diff
@@ -0,0 +1,58 @@
+// @ts-expect-error
+const { kakao } = window;
+
+export const useKakaoMap = (
```
이야.. 고생 많으셨네요..👍👍👍👍👍👍

geongyu09 commented on src/hooks/usePlaceSearch.ts:
```diff
@@ -0,0 +1,40 @@
+
+import type { PlaceInfo } from '@/types';
+
+const KAKAO_API_KEY = 'fe995452b0e63ba6e5276669fad15c21';
```
혹시 해당 키값은 올려도 괜찮은 값인가요? 
잘못하면 큰 문제가 생길 수도 있을 것 같습니다..!

Da-Yeon-Kim replied:
아, 이거는 안 그래도 추후에 .env 파일에 따로 작성해서 빼두었습니다! .env 파일은 git으로 공유가 안 되는데 어떻게 서로 소통해야 할까요?? 카톡으로... 해야 할까요?

geongyu09 commented on src/hooks/usePlaceSearch.ts:
```diff
@@ -0,0 +1,40 @@
+    const searchPlaceByKeyword = async (keyword: string) => {
+      const url = `https://dapi.kakao.com/v2/local/search/keyword.json?query=${keyword}`;
+
+      const response = await fetch(url, {
```
혹시 여기서 fetch api를 사용한 이유가 있나요?? 

기존에 저희가 정의했던 게층구조와 맞지 않는 느낌이 듭니다!

Da-Yeon-Kim replied:
앗, 그렇네요!
카카오맵 API를 사용하는 로직을 `hooks/` 폴더 내에 다 우겨넣다 보니 이렇게 됐네요ㅋㅋㅋ
계층 구조에 맞추어 수정하겠습니다! 👍

geongyu09 commented on src/hooks/useKakaoMap.ts:
```diff
@@ -0,0 +1,58 @@
+
+    const marker = new kakao.maps.Marker({
+      position: map.getCenter(),
+      map: map,
```
map: map, 은 map, 으로 변경해도 좋아보입니다!

geongyu09 commented on src/hooks/useKakaoMap.ts:
```diff
@@ -0,0 +1,58 @@
+  useEffect(() => {
+    if (!defaultPosition) return;
+
+    const container = document.getElementById(containerId);
```
ref를 사용해도 좋아보입니다만.. 수정할 부분이 많아질 수 있어서..
그냥 둬도 좋아보입니다!

geongyu09 commented on src/components/features/Create/CreateMap/index.tsx:
```diff
@@ -1,16 +1,40 @@
+    }
+  }, [placeInfo, setValue, coordinates?.lat, coordinates?.lng]);
+
+  return <MapContainer id="map" />;
```
id를 "map"으로 작성해도 좋지만, useRef나 useId를 사용해도 좋아보입니다. 
그게 아니라면, 상수로 관리해서, 나중에 다른 곳에서 사용할 때 휴먼 에러를 줄일 수 있을 것 같습니다!
여기서는 coordinates() 의 첫 인자에 "map" 리터럴이 그대로 들어가니.. 수정시에도 같이 수정을 해줘야 하므로 상수로 관리하는 것이 좋아보입니다!

Da-Yeon-Kim committed:
refactor: createMeeting 요청, 응답 데이터 분리해 관리

- types 폴더에서 CreateMeetingRequest, Response 타입 관리
- useCreateFormContext에서도 요청 데이터와 타입 같도록 조정
- API_KEY는 .env 파일에 추가하여 보안 관리

 #50

geongyu09 reviewed (approved):
고생하셨습니다! 정말로 고생하셨어요..
엄청 어려운 부분이었는데, 이렇게 잘 만들어 주시다니.. 저도 잘 가져다 사용하겠습니다

물론 리액트스럽지 않는 코드가 존재하긴 하는 것 같습니다..! 다만.. 이를 수정하기에는 너무나도 많은 비용이 들 것 같아서.. 
지금은 잘 돌아가는 코드이니, 코드 수정은 나중에 해도 좋지 않을까 합니다! 

---

추가적으로 지도가 보이기 전에 아무런 화면이 보이지 않는데, default 화면을 만들어 봐도 좋아보입니다! 

진짜 수고많으셨습니다!

Da-Yeon-Kim committed:
feat: startDate, endDate를 Date 객체로 변경

 #50

HyeonChae1104 reviewed (approved):
헉!! 너무 잘 만드셨어요..ㅜ
고생 많으셨습니다!!! 저는 내 현재 위치가 잘 불러와지네요!! 
생성페이지 어려운 부분 많으셨을텐데 수고하셨습니다!ㅎㅎ

Da-Yeon-Kim commented:
제가 막 타입을 통일해서 사용해 보려다가 현재는 오류가 발생하고 있어서,,
이 이슈 해결하고, PR 리뷰 받은 수정사항들 조금씩 바꾸고 난 다음에 머지하겠습니다!

Da-Yeon-Kim committed:
feat: Date 객체 string으로 변환

- Date나 Time보다는 string으로 관리하는 게 편해 보여서 다시
  바꾸었습니다....
  - kakao local api를 사용하는 코드의 위치를 /api/hooks로 이동
  - validation 빼먹은 ! 추가
  - PR 리뷰 간단한 몇 가지 반영 (map, 구조 분해 할당 등)

   #50
